### PR TITLE
Don't regain breath while in ignore node

### LINKED ISSUE #8217: Player regains a little breath when rejoining game into ignore
##### Issue type
 - Bug report
 
##### Minetest version
```
Minetest 5.0.0-dev-a809f736
```

##### OS / Hardware

Operating system: Ubuntu 18.10
CPU: AMD Ryzen 7 2700

##### Summary
(When the background is grey I've just joined the game)
water

##### Steps to reproduce
Leave the game while underwater and half out of breath. Then rejoin and watch your bubble meter. Might need to enable some heavy mods

### PR DESCRIPTION
**Pull request by Wuzzy2**
Fixes #8217.

The player shall neither lose or gain breath while in `ignore` because we don't know yet if the node will turn out to be a drowning node or a non-drowning one.

How to test: Start simple server, can even be locally from client. Drop in liquid. Lose breath. Stop server. Restart server. Look at breath meter. If the breath meter does not change, it works.

### CODE REVIEW

**Comment by Wuzzy2** [comment]
Oops. I meant to say: “If the breath meter does not *increase* before the liquid has loaded, it works.”

File: src/content_sao.cpp [hunk 1]
```diff
@@ -1015,9 +1015,10 @@ void PlayerSAO::step(float dtime, bool send_recommended)
 		// Get nose/mouth position, approximate with eye position
 		v3s16 p = floatToInt(getEyePosition(), BS);
 		MapNode n = m_env->getMap().getNodeNoEx(p);
+		content_t content = n.getContent();
 		const ContentFeatures &c = m_env->getGameDef()->ndef()->get(n);
-		// If player is alive & no drowning, breathe
-		if (m_hp > 0 && m_breath < m_prop.breath_max && c.drowning == 0)
+		// If player is alive & no drowning & not in ignore, breathe
+		if (content != CONTENT_IGNORE && m_hp > 0 && m_breath < m_prop.breath_max && c.drowning == 0)
```
Comment (SmallJoker) [inline src/content_sao.cpp]: Line is too long, please break.

**Comment by Wuzzy2** [comment]
@Krock: OK, done.

File: src/content_sao.cpp [hunk 2]
```diff
@@ -1015,9 +1015,11 @@ void PlayerSAO::step(float dtime, bool send_recommended)
 		// Get nose/mouth position, approximate with eye position
 		v3s16 p = floatToInt(getEyePosition(), BS);
 		MapNode n = m_env->getMap().getNodeNoEx(p);
+		content_t content = n.getContent();
 		const ContentFeatures &c = m_env->getGameDef()->ndef()->get(n);
-		// If player is alive & no drowning, breathe
-		if (m_hp > 0 && m_breath < m_prop.breath_max && c.drowning == 0)
+		// If player is alive & no drowning & not in ignore, breathe
+		if (content != CONTENT_IGNORE && m_hp > 0 &&
```
Comment (paramat) [inline src/content_sao.cpp]: Can this line be re-arranged to short-circuit as early as possible? So terms most likely to be false earliest.
Not being in ignore is probably the least likely to be false so should probably be last.
Best order seems ot be:
m_breath < m_prop.breath_max
c.drowning == 0
content != CONTENT_IGNORE
m_hp > 0

**Comment by Wuzzy2** [comment]
@paramat: Look again, please. :-)

File: src/content_sao.cpp (hunk 2 above)
Comment (paramat) [reply src/content_sao.cpp]: Done thanks.

**Comment by paramat** [comment]
Code looks good now, has anyone tested it?

**Comment by paramat** [comment]
:+1: but a minor request if you don't mind:
```
		if (m_breath < m_prop.breath_max &&
				c.drowning == 0 && n.getContent() != CONTENT_IGNORE && m_hp > 0)
```
Then there's no need to create an extra variable 'content' at line 1018.

**Comment by paramat** [comment]
Yes that's good because then `n.getContent()` doesn't get executed unless it needs to be.

**Pull request merged**